custom_components/xiaomi_miio_fan/fan.py

```diff
@@ -1748,3 +1792,346 @@ def delay_off(self, seconds: int):
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of the fan."""
+        _LOGGER.debug("Setting the preset mode to: %s", preset_mode)
+
+        if not self._state:
+            await self._try_command(
+                "Turning the miio device on failed.", self._device.on
+            )
+        await self._try_command(
+            "Setting preset mode of the miio device failed.",
+            self._device.set_speed,
+            FAN_PRESET_MODE_VALUES[preset_mode],
+        )
+
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed percentage of the fan."""
+        _LOGGER.debug("Setting the fan speed percentage to: %s", percentage)
+
+        if percentage == 0:
+            await self.async_turn_off()
+            return
+
+        if not self._state:
+            await self._try_command(
+                "Turning the miio device on failed.", self._device.on
+            )
+        await self._try_command(
+            "Setting preset mode of the miio device failed.",
+            self._device.set_speed,
+            percentage,
+        )
+
+    async def async_oscillate(self, oscillating: bool) -> None:
+        """Set oscillation."""
+        if oscillating:
+            await self._try_command(
+                "Setting oscillate on of the miio device failed.",
+                self._device.set_oscillate,
+                True,
+            )
+        else:
+            await self._try_command(
+                "Setting oscillate off of the miio device failed.",
+                self._device.set_oscillate,
+                False,
+            )
+
+    async def async_set_delay_off(self, delay_off_countdown: int) -> None:
+        """Set scheduled off timer in minutes."""
+
+        await self._try_command(
+            "Setting delay off miio device failed.",
+            self._device.delay_off,
+            delay_off_countdown * 60,
+        )
+
+    async def async_set_natural_mode_on(self):
+        """Turn the natural mode on."""
+        if self._device_features & FEATURE_SET_NATURAL_MODE == 0:
+            return
+
+        await self._try_command(
+            "Setting fan natural mode of the miio device failed.",
+            self._device.set_mode,
+            OperationModeFanP33.NaturalWind,
+        )
+
+    async def async_set_natural_mode_off(self):
+        """Turn the natural mode off."""
+        if self._device_features & FEATURE_SET_NATURAL_MODE == 0:
+            return
+
+        await self._try_command(
+            "Setting fan natural mode of the miio device failed.",
+            self._device.set_mode,
+            OperationModeFanP33.StraightWind,
+        )
+
+
+class OperationModeFanP33(Enum):
+    StraightWind = 0
+    NaturalWind = 1
+
+class FanStatusP33(DeviceStatus):
+    """Container for status reports for FanP33."""
+
+    def __init__(self, data: Dict[str, Any]) -> None:
+        self.data = data
```

Comment:
Could you add a docstring containing a example response?

Reply:
Sorry, I am not very well-versed with python so I am trying to understand your request a bit better. You'd like a docstring for `__init__` showing what the structure of `data` would look like?

Reply:
Yes. I would like to see something like this: https://github.com/rytilahti/python-miio/blob/master/miio/integrations/fan/zhimi/fan.py#L56-L76

Reply:
Ok sounds good. I have been trying to print the data to the log, but the log doesn't seem to appear on the HASS log. The only way I get something is if I add the custom component to the `configuration.yaml` `logger:`, and it returns something like

```
<FanStatusP33 buzzer=True child_lock=None fan_level=2 fan_speed=59 led=True light=True mode=OperationModeFanP33.StraightWind power=True power_off_time=0 speed_rpm=59 swing_mode=False swing_mode_angle=140 temperature=KeyError>
```

I am pretty sure I don't have a correctly setup dev environment, so perhaps I am doing something silly here. 

To get the data from `__init__` I am adding this to the `status()` function

```
fan_status = FanStatusP33(
            {
                prop["did"]: prop["value"] if prop["code"] == 0 else None
                for prop in self.get_properties_for_mapping()
            }
        )
        _LOGGER.debug("TEST TEST Data: %s", fan_status)
```

and then calling it from something like `set_mode()`. When I set the mode vcia the integration, _LOGGER doesn't print to the log

Reply:
@syssi printing `data` returns something like this, which is not quite the same format as the examples you gave

```
{'id': 7, 'result': [{'did': 'power', 'siid': 2, 'piid': 1, 'code': 0, 'value': True}, {'did': 'fan_level', 'siid': 2, 'piid': 2, 'code': 0, 'value': 1}, {'did': 'oscillate', 'siid': 2, 'piid': 4, 'code': 0, 'value': False}, {'did': 'angle', 'siid': 2, 'piid': 5, 'code': 0, 'value': 120}, {'did': 'mode', 'siid': 2, 'piid': 3, 'code': 0, 'value': 1}, {'did': 'delay_off_countdown', 'siid': 3, 'piid': 1, 'code': 0, 'value': 0}, {'did': 'child_lock', 'siid': 7, 'piid': 1, 'code': 0, 'value': False}, {'did': 'light', 'siid': 4, 'piid': 1, 'code': 0, 'value': True}, {'did': 'buzzer', 'siid': 5, 'piid': 1, 'code': 0, 'value': True}, {'did': 'motor_control', 'siid': 6, 'piid': 1, 'code': -4003}, {'did': 'speed', 'siid': 2, 'piid': 6, 'code': 0, 'value': 20}], 'exe_time': 150}
```

Should I just infer from this result, and manually write the docstring, or is there a better way of doing this?

Reply:
I would suggest to log the data directly at the status container to retrieve the exact contents of the variable. You could use `_LOGGER.error()` if you want to make sure it gets logged.

Reply:
The output I pasted above is directly from the status function

```
def status(self):
        """Retrieve properties."""
        status = FanStatusP33(
            {
                prop["did"]: prop["value"] if prop["code"] == 0 else None
                for prop in self.get_properties_for_mapping()
            }
        )
        _LOGGER.debug("MIIO STATUS: %s", status.data)
        return status
```

The formatting of the example you provided is different though

```
Response of a Fan (zhimi.fan.sa4):
        {'angle': 120, 'speed': 327, 'poweroff_time': 0, 'power': 'on',
         'ac_power': 'on', 'angle_enable': 'off', 'speed_level': 1, 'natural_level': 0,
         'child_lock': 'off', 'buzzer': 2, 'led_b': 0, 'use_time': 85}
"""

Reply:
I'm looking for the raw data passed to the class / you could log the return value of

```
                prop["did"]: prop["value"] if prop["code"] == 0 else None
                for prop in self.get_properties_for_mapping()
```

The return value of `FanStatusP33(data)` is the interpretation of the raw data.

Reply:
Uuuh. You are right. I was confused. This content is fine:

```
{'id': 7, 'result': [{'did': 'power', 'siid': 2, 'piid': 1, 'code': 0, 'value': True}, {'did': 'fan_level', 'siid': 2, 'piid': 2, 
'code': 0, 'value': 1}, {'did': 'oscillate', 'siid': 2, 'piid': 4, 'code': 0, 'value': False}, {'did': 'angle', 'siid': 2, 'piid': 5, 
'code': 0, 'value': 120}, {'did': 'mode', 'siid': 2, 'piid': 3, 'code': 0, 'value': 1}, {'did': 'delay_off_countdown', 'siid': 3, 
'piid': 1, 'code': 0, 'value': 0}, {'did': 'child_lock', 'siid': 7, 'piid': 1, 'code': 0, 'value': False}, {'did': 'light', 'siid': 4,
 'piid': 1, 'code': 0, 'value': True}, {'did': 'buzzer', 'siid': 5, 'piid': 1, 'code': 0, 'value': True}, {'did': 'motor_control', 
'siid': 6, 'piid': 1, 'code': -4003}, {'did': 'speed', 'siid': 2, 'piid': 6, 'code': 0, 'value': 20}], 'exe_time': 150}
```

Please add it mentioning the model name and firmware version number as docstring.

Reply:
Sounds good

Reply:
Another **miot** device response as example: https://github.com/rytilahti/python-miio/blob/master/miio/integrations/fan/zhimi/zhimi_miot.py#L50-L72